vulcan/lib/client/jsx/components/workflow/user_interactions/inputs/data_transformation.tsx

```diff
@@ -0,0 +1,219 @@
+
+import {joinNesting} from './monoids';
+import {WithInputParams, DataEnvelope} from './input_types';
+import {useSetsDefault} from './useSetsDefault';
+import {some, Maybe} from '../../../../selectors/maybe';
+import {useMemoized} from '../../../../selectors/workflow_selectors';
+
+const useStyles = makeStyles((theme) => ({
+  dialog: {
+    maxWidth: '90vw',
+    maxHeight: '90vh'
+  }
+}));
+
+function DataTransformationModal({
+  data,
+  onChange,
+  onClose
+}: {
+  data: any[][];
+  onChange: (data: Maybe<{[key: string]: any}>) => void;
+  onClose: () => void;
+}) {
+  const classes = useStyles();
+
+  const hotTableComponent = useRef<any>(null);
+  const hyperformulaInstance = HyperFormula.buildEmpty({
+    // to use an external HyperFormula instance,
+    // initialize it with the `'internal-use-in-handsontable'` license key
+    licenseKey: 'internal-use-in-handsontable'
+  });
+
+  return (
+    <>
+      <DialogTitle>Transform your data</DialogTitle>
+      <DialogContent className={classes.dialog}>
+        <HotTable
+          ref={hotTableComponent}
+          settings={{
+            data: data,
+            colHeaders: true,
+            rowHeaders: true,
+            height: 'auto',
+            licenseKey: 'non-commercial-and-evaluation',
+            formulas: {
+              engine: hyperformulaInstance
+            },
+            contextMenu: {
+              items: {
+                col_left: {
+                  name: 'Insert column to left'
+                },
+                col_right: {
+                  name: 'Insert column to right'
+                },
+                remove_col: {},
+                undo: {},
+                redo: {},
+                clear_column: {}
+              }
+            }
+          }}
+        />
+      </DialogContent>
+      <DialogActions>
+        <Button
+          onClick={onClose}
+          startIcon={<CancelIcon />}
+          color='secondary'
+          variant='contained'
+        >
+          Cancel
+        </Button>
+        <Button
+          onClick={() => {
+            if (hotTableComponent.current) {
+              // hotTableComponent.current.hotInstance.getSourceData()
+              //   returns the raw formulas, unrendered, which we
+              //   set as input values to preserve any transformations.
+              const sourceData = hotTableComponent.current.hotInstance.getSourceData();
+              const data = hotTableComponent.current.hotInstance.getData();
+
+              onChange(
+                some({
+                  source_data: toJson(sourceData),
+                  data: toJson(data)
+                })
+              );
+            }
+
+            onClose();
+          }}
+          startIcon={<SaveIcon />}
+          color='primary'
+          variant='contained'
+        >
+          Save
+        </Button>
+      </DialogActions>
+    </>
+  );
+}
+
+export function toJson(input: any[][]): DataEnvelope<{[key: string]: any}> {
```

Comment:
Should this be exported or just local module function?  Seems fairly specific to this component and has a very generic name.

Reply:
Exported for tests only, basically...

Reply:
I can rename it to be more specific. I was too lazy to write a component-level test since the table opens in a dialog, so just wanted to test the core transformation logic.